### PR TITLE
fix(signature): drop empty signed Claude thinking

### LINKED ISSUE #3663: Sanitizer leaves signed thinking blocks with empty text intact, causing Anthropic    400 "thinking blocks cannot be modified"
**Is it a request payload issue?**
[x] Yes, this is a request payload issue. I am using a client/cURL to send a request payload, but I received an unexpected error.
[  ] No, it's another issue.

**If it's a request payload issue, you MUST know**
Our team doesn't have any GODs or ORACLEs or MIND READERs. Please make sure to attach the request log or curl payload.

See below. I have a fix.

**Describe the bug**

Summary                                                                                 

When a client round-trips an assistant message and strips the thinking text but keeps the signature field, the Claude-upstream sanitizer (SanitizeClaudeMessagesForClaudeUpstream) preserves the block unchanged. Anthropic then rejects the request with:        
                                                  
`400 invalid_request_error: messages.X.content.Y: `thinking` or `redacted_thinking` `

blocks in the latest assistant message cannot be modified. These blocks must remain as  they were in the original response.   


**CLI Type**
What type of CLI account do you use?  (gemini-cli, gemini, codex, claude code or openai-compatibility)

`opencode` with Anthropic Opus 4.8 Thinking

**Model Name**

Opus 4.8 Thinking

**LLM Client**

`opencode` 

**Request Information**

Repro                                                                                   
                                                                                      
Send /v1/messages to a Claude upstream with an assistant turn containing:             
  
{"type": "thinking", "thinking": "", "signature": "<valid-looking Claude E-prefix signature ~1KB>"}                                                                      

Observed with opencode 2.x using @ai-sdk/anthropic against claude-opus-4-8(xhigh) after a tool-use turn.                                                                        
                                                                                     
                                                                       
                 
 In internal/signature/claude_messages_sanitize.go::SanitizeClaudeMessagesSignaturesForTarget, the loop checks isEmptyClaudeThinkingPlaceholder (which requires both empty signature AND empty text), then jumps to DecideSignatureCompatibility which sees a  valid-shaped Claude signature and returns SignatureActionPreserve. The asymmetric case -  signature present + text empty — is never caught, so the broken block is forwarded as-is. Anthropic's server-side signature validation then fails because the signature does not validate against empty text.


**Expected behavior**

```
Patch (against 3a54fb7f):                                                               
  rawSignature := part.Get("signature").String()
  // A signed thinking block whose text has been stripped (e.g. by a                      
  // client that round-trips assistant messages and drops the thinking
  // payload) will fail Anthropic's signature-vs-text validation.                         
  // Drop the entire block; Anthropic accepts missing thinking blocks.
  if strings.TrimSpace(rawSignature) != "" &&                                             
  strings.TrimSpace(claudeThinkingBlockText(part)) == "" {                                
         report.DroppedBlocks++                                                              
          messageModified = true                                                              
         continue  
  }
 decision := DecideSignatureCompatibility(targetProvider, rawSignature, 
  SignatureBlockKindClaudeThinking)        
```                                               
 
Plus a test (matches existing test style in `provider_compatibility_test.go`).            



**OS Type**
 - OS: Linux x86 
 - Version Ubuntu 24.04 LTS
 - CLIProxyAPI version from Git

**Additional context**

Running this patch locally for ~2 days against opus-4-8 with opencode; ~0 occurrences of the 400 since. Happy to open a PR with the fix + test if welcome.

### PR DESCRIPTION
**Pull request by wilsonicdev**
Fixes #3663

## Summary
- drop Claude `thinking` blocks when a non-empty Claude-target signature is replayed with empty/stripped thinking text
- keep the rest of the message content intact so conversations can continue instead of forwarding an invalid signed block upstream
- add a regression test for a valid-looking Claude signature paired with empty thinking text
- preserve GPT/Codex signature-only reasoning blocks when the target provider is GPT/Codex

## Rationale
Claude validates signed thinking blocks against their original thinking text. If a client round-trips an assistant message and strips the text but keeps the signature, preserving that block for a native Claude upstream can trigger a provider-side signature validation error. Dropping the invalid Claude-target block matches the existing sanitizer behavior for incompatible thinking blocks and avoids the 400.

## Review follow-up
- addressed Codex automated review feedback by limiting the empty-text drop to Claude targets
- added a regression test to ensure GPT/Codex signature-only reasoning remains preserved for GPT targets

## Validation
Ran inside `golang:1.26-alpine` because the host has no Go toolchain installed:

- `/usr/local/go/bin/gofmt -w internal/signature/claude_messages_sanitize.go internal/signature/provider_compatibility_test.go`
- `/usr/local/go/bin/go test ./internal/signature -run "TestSanitizeClaudeMessagesForClaudeUpstream_DropsSignedThinkingWithEmptyText|TestSanitizeClaudeMessagesSignaturesForModel_PreservesGPTSignatureOnlyReasoningForGPT" -count=1`
- `/usr/local/go/bin/go test ./internal/signature -count=1`
- `/usr/local/go/bin/go test ./... -count=1`
- `/usr/local/go/bin/go build -o test-output ./cmd/server && rm test-output`

### CODE REVIEW

File: internal/signature/claude_messages_sanitize.go [hunk 1]
```diff
@@ -124,6 +124,15 @@ func SanitizeClaudeMessagesSignaturesForTarget(payload []byte, opts ClaudeMessag
...
 			rawSignature := part.Get("signature").String()
+			// A signed thinking block whose text was stripped by a client cannot be
+			// replayed safely: provider-side signature validation checks the
+			// signature against the original thinking text. Drop the block so the
+			// conversation can continue without replaying invalid signed content.
+			if strings.TrimSpace(rawSignature) != "" && strings.TrimSpace(claudeThinkingBlockText(part)) == "" {
+				report.DroppedBlocks++
+				messageModified = true
+				continue
+			}
```
[the remark this answers is not in the document]
Comment (wilsonicdev) [reply internal/signature/claude_messages_sanitize.go]: Fixed in fc38c717: the empty-text signed-thinking drop is now limited to Claude targets, so GPT/Codex signature-only reasoning is preserved. I also added `TestSanitizeClaudeMessagesSignaturesForModel_PreservesGPTSignatureOnlyReasoningForGPT` and re-ran the focused tests, `go test ./internal/signature`, `go test ./...`, and the server build.

**Comment by wilsonicdev** [comment]
@codex review

Follow-up was pushed in `fc38c717` after the previous review: the empty signed-thinking drop is now limited to Claude targets, and GPT/Codex signature-only reasoning has a regression test to ensure it remains preserved. Please review the current head commit.
